enhancements/two-nodes-openshift/2no.md

```diff
@@ -0,0 +1,398 @@
+
+In the case of the core OpenShift installer, the user-facing proceedure is unchanged from a standard "IPI" installation, other than the configuration of 2 nodes instead of 3.
+Internally, the RedFish details for each node will need to make their way into the RHEL-HA configuration, but this is information already required for bare-metal hosts.
+
+In the case of the Assisted Installer, the user-facing proceedure follows the standard flow except for the configuration of 2 nodes instead of 3, and the collection of RedFish details (including passwords!) for each node which are needed for the RHEL-HA configuration.
+
+Everything else about cluster creation will be an opaque implementation detail not exposed to the user. 
+
+#### Day 2 Proceedures
+
+As per a standard 3-node control-plane, OpenShift upgrades and `MachineConfig` changes can not be applied when the cluster is in a degraded state.
+Such operations will only proceed when both peers are online and healthy.
+
+The experience of managing a 2-node control-plane should be largely indistinguishable from that of a 3-node one.
+The primary exception is (re)booting one of the peers while the other is offline, and expected to remain so.
+
+As in a 3-node control-plane cluster, starting only one node is not expected to result in a functioning cluster.
+Should the admin wish for the control-plane to start, the admin will need to execute a supplied confirmation command on the active cluster node. 
+This command will grant quorum to the RHEL-HA components, authorizing it to fence it's peer and start etcd in as a cluster-of-one read/write mode.
+Confirmation can be given at any point and optionally make use of SSH to facilitate initiation by an external script.
+
+### API Extensions
+
+There are two related but ultimately orthogonal capabilities that may require API extensions.
+
+1. Identify the cluster as having a unique topology
+2. Tell CEO when it is safe for it to disable certain membership related functionalities
+
+#### Unique Topology
+
+A mechanism is needed for components of the cluster to understand that this is a 2 node control-plane topology which may require different handling.
+We will define a new value for the `TopologyMode` enum: `DualReplicaTopologyMode`.
+
+However `TopologyMode` is not available at the point the Agent Based Installer (ABI) performs validation.
+We will therefore additionally define a new feature gate `DualReplicaTopology` that can be enabled in `install-config.yaml`, and which ABI can use to validate the proposed cluster - such as the proposed node count.
+
+#### CEO Trigger
+
+Initially the creation of an etcd cluster will be driven in the same way as other platforms.
+Once the cluster has two members, the etcd daemon will be removed from the static pod definition and recreated as a resource controlled by RHEL-HA.
```

Comment:
Is the plan for your startup job to busy-wait until it can poll from the API that the cluster has 2 members? Then it tries to update the CEO CR, which has etcd dropped from CEO. We busy-wait until we see that in the CEO CR, and then we proceed with the update.

IMHO, this sounds like it would be better served by an operator - since this is a reconciliation loop. One concern I have is what do we do if the systemd job fails for some unexpected reason? What is going to reconcile that? In my experience, the expectation we have of the installer is that it will set up what's needed to get the cluster going, set up the initial operators (this might actually be CVO moreso than the installer), and *wait* for those operators to reconcile the desired state.

It feels to me like what we want is an operator that can own the reconciliation steps for taking a "normal" 2node cluster and transforming it into one where pacemaker and corosync are running and etcd is run by this cnf script instead of CEO.

Reply:
My assumption had been for a startup job and a busy-wait loop, but I think most (all?) are thinking an operator-like* entity is the way to go now.
 
* "operator-like" since it needs to remain functional even when etcd is offline or inquorate

Reply:
Note that etcd deployment goes through multiple (~8) revisions and rollouts during the installation (mainly get updated with different crts and CAs).
Also CEO is the one that knows what the etcd member deployment should look like, how do you plan to deploy etcd without CEO?

Reply:
@eranco74 you can assume that CEO will finish its gazillion rollouts and then the switch happens. So it's a fully finished and rendered yaml.

The certificates and their rotation are a good point, which we need to resolve for GA. Currently the CEO would just replace the files on disk through a revision installation, but the podman container running etcd would not restart accordingly.

Reply:
we might add a check in etcd pacemaker agent for refreshed certificates and reboot the etcd instance